### PR TITLE
Add OMR_SOCKET_API flag to enable/disable usage of socket API

### PR DESCRIPTION
**Pull request by caohaley**
Created the flag for a more lightweight OMR library if user prefers

- Created the flag and added it to build systems including autoconf and
  CMakeo.

Issue: #4102

Co-authored-by: Babneet Singh <sbabneet@ca.ibm.com>

Signed-off-by: Haley Cao <haleycao88@hotmail.com>

### CODE REVIEW

**Comment by caohaley** [comment]
@rwy0717 Should I add this flag for the socket API? This flag enables and disables the usage of the socket API.

**Comment by rwy0717** [comment]
Yep, this seems like a good idea to me :+1: Although I'm not sure the flag name is right, maybe `OMR_PORT_SOCKET_SUPPORT`? We'll also want to turn this flag on in CI (by enabling it in cmake/caches/Travis.cmake) to make sure we're testing the new code.

**Comment by rwy0717** [comment]
@genie-omr build all

**Comment by rwy0717** [comment]
aarch failed due to infrastructure issues.
@genie-omr build aarch64

**Comment by caohaley** [comment]
I modified Travis.make to `set(OMR_PORT_SOCKET_SUPPORT ON CACHE BOOL "")` to enable socket api, but I'm don't think the socket api code was compiled in the PR checks (in `continuous-integration/appveyor/pr`). I think I may need to force set in Travis.make to successfully enable the flag for the CI (`set(OMR_PORT_SOCKET_SUPPORT ON CACHE BOOL "") FORCE`).

**Comment by rwy0717** [comment]
@genie-omr build xlinux

**Comment by rwy0717** [comment]
Hi Hailey, I built and ran the tests locally and confirmed that the tests do run when the flag is enabled.

The test also ran in the xlinux build above, look for:
```
[2019-12-03T19:55:58.115Z] 31: [----------] 4 tests from PortSockTest
[2019-12-03T19:55:58.115Z] 31: [----------] 4 tests from PortSockTest (0 ms total)
```

What I think went wrong for you:

Travis.cmake is an initial cache file, which will only initialize variables the first time cmake configure a build directory. After that first run, variable values are pulled from CMakeCache.txt in the build directory. So if you're trying to update the configuration for an existing build, editing Travis.cmake will not be enough. You have to forcibly set a variable from the command line by doing:

```
cd build/
cmake -DOMR_SOCKET_SUPPORT=ON
```

**Comment by caohaley** [comment]
@babsingh

**Comment by babsingh** [comment]
You also need to update https://github.com/eclipse/omr/blob/master/cmake/caches/Appveyor.cmake; it gets used by the Appveyor builds. @caohaley Does this change fix your issues?

File: port/omrportpriv.h [hunk 1]
```diff
@@ -635,6 +636,7 @@ extern J9_CFUNC int32_t
 omrsock_recvfrom(struct OMRPortLibrary *portLibrary, omrsock_socket_t sock, uint8_t *buf, int32_t nbyte, int32_t flags, omrsock_sockaddr_t addrHandle);
 extern J9_CFUNC int32_t
 omrsock_close(struct OMRPortLibrary *portLibrary, omrsock_socket_t *sock);
+#endif /* OMR_PORT_SOCKET_SUPPORT */
```
Comment (babsingh) [inline port/omrportpriv.h]:
```suggestion
#endif /* defined(OMR_PORT_SOCKET_SUPPORT) */
```

File: port/common/omrport.c [hunk 2]
```diff
@@ -316,6 +317,7 @@ static OMRPortLibrary MasterPortLibraryTable = {
 	omrsock_recv, /* sock_recv */
 	omrsock_recvfrom, /* sock_recvfrom */
 	omrsock_close, /* sock_close */
+#endif /* OMR_PORT_SOCKET_SUPPORT */
```
Comment (babsingh) [inline port/common/omrport.c]:
```suggestion
#endif /* defined(OMR_PORT_SOCKET_SUPPORT) */
```

File: include_core/omrport.h [hunk 3]
```diff
@@ -1720,6 +1721,8 @@ typedef struct OMRPortLibrary {
 	int32_t (*sock_recvfrom)(struct OMRPortLibrary *portLibrary, omrsock_socket_t sock, uint8_t *buf, int32_t nbyte, int32_t flags, omrsock_sockaddr_t addrHandle) ;
 	/** see @ref omrsock.c::omrsock_close "omrsock_close"*/
 	int32_t (*sock_close)(struct OMRPortLibrary *portLibrary, omrsock_socket_t *sock) ;
+#endif /* OMR_PORT_SOCKET_SUPPORT */
```
Comment (babsingh) [inline include_core/omrport.h]:
```suggestion
#endif /* defined(OMR_PORT_SOCKET_SUPPORT) */
```

File: include_core/omrport.h [hunk 4]
```diff
@@ -2177,6 +2182,7 @@ extern J9_CFUNC int32_t omrport_getVersion(struct OMRPortLibrary *portLibrary);
 #define omrsock_recv(param1,param2,param3,param4) privateOmrPortLibrary->sock_recv(privateOmrPortLibrary, (param1), (param2), (param3), (param4))
 #define omrsock_recvfrom(param1,param2,param3,param4,param5) privateOmrPortLibrary->sock_recvfrom(privateOmrPortLibrary, (param1), (param2), (param3), (param4), (param5))
 #define omrsock_close(param1) privateOmrPortLibrary->sock_close(privateOmrPortLibrary, (param1))
+#endif /* OMR_PORT_SOCKET_SUPPORT */
```
Comment (babsingh) [inline include_core/omrport.h]:
```suggestion
#endif /* defined(OMR_PORT_SOCKET_SUPPORT) */
```

**Comment by rwy0717** [comment]
@genie-omr build all

**Comment by caohaley** [comment]
@babsingh  @rwy0717  I tested with Jenkins, and they passed. Ready for review. Seems like there is a conflicting file, should I try to resolve it and push it again? May need to run all the tests again in that case.

**Comment by rwy0717** [comment]
Yeah you'll have to resolve the conflict 👍

**Comment by babsingh** [comment]
> May need to run all the tests again in that case.

No need to run the tests. Just make sure that your changes are preserved after the conflict.

**Comment by rwy0717** [comment]
running anyways :)
@genie-omr build all

**Comment by rwy0717** [comment]
@genie-omr build aix

**Pull request merged**